### ISSUE TITLE
Pileup browser doesn't perform well when files are stored locally in the machine and not on a server

### ISSUE THREAD

**Issue by ifll** [hammerlab/pileup.js#436]
Recently I have been trying to add pileup.js browser to an application I am building within Django framework 1.9.8.

This application requires to be displayed "reference", "scale", "location", "coverage" and "alignments".

I have made many trials to create the pileup browser described below:

- **_T1_**: reference (hg38.2bit), .bam and .bai files (bodymap-skeletal_muscle-chr22) retrieved from "http://www.biodalliance.org/datasets/"
In this case the application loads very fast and data is available with no problems. I have checked the ajax queries sent to retrieve the data and they come in pairs, retrieving only partial content.

**JAVASCRIPT CODE:**
```javascript
window.divGB = document.getElementById('genomeBrowserHolder');
window.range = {contig: 'chr17', start: 7512384, stop: 7512544};
window.bamSource = pileup.formats.bam({
    	url: "http://www.biodalliance.org/datasets/bodymap-skeletal_muscle-chr22.bam",
        indexUrl: "http://www.biodalliance.org/datasets/bodymap-skeletal_muscle-chr22.bam.bai"
	});
window.sources = [
	  {
	      viz: pileup.viz.genome(),
	      isReference: true,
	      data: pileup.formats.twoBit({
	      	url: 'http://www.biodalliance.org/datasets/hg38.2bit'
	      }),
	      name: 'Reference'
	  },
	  {
	      viz: pileup.viz.scale(),
	      name: 'Scale'
	  },
	  {
	      viz: pileup.viz.location(),
	      name: 'Location'
	  },
	  {
	      viz: pileup.viz.coverage(),
	      data: bamSource,
	      name: 'Coverage'
	  },
	  {
	      viz: pileup.viz.pileup({
	        viewAsPairs: true
	      }),
	      data: bamSource,
	      cssClass: 'normal',
	      name: 'Alignments'
	  }
	];
window.p = pileup.create(divGB, {
		  range: range,
		  tracks: sources
		});
```

- **_T2:_** reference (hg38.2bit) from biodallience and .bam and .bai files (synth3.normal.17.7500000-7515000) from local file in the computer. The reference loads as before (in pieces and very fast) and the bam and bai files are completely loaded, and, as they weight 1,1MB toghether, they are loaded fast.

**JAVASCRIPT CODE:**
```javascript
window.divGB = document.getElementById('genomeBrowserHolder');
window.range = {contig: 'chr17', start: 7512384, stop: 7512544};
window.bamSource = pileup.formats.bam({
    	url: "{{MEDIA_URL}}uploads/{{project_id}}/bwa/synth3.normal.17.7500000-7515000.bam",
        indexUrl: "{{MEDIA_URL}}uploads/{{project_id}}/bwa/synth3.normal.17.7500000-7515000.bam.bai"
	});
window.sources = [
	  {
	      viz: pileup.viz.genome(),
	      isReference: true,
	      data: pileup.formats.twoBit({
	      	url: 'http://www.biodalliance.org/datasets/hg38.2bit'
	      }),
	      name: 'Reference'
	  },
	  {
	      viz: pileup.viz.scale(),
	      name: 'Scale'
	  },
	  {
	      viz: pileup.viz.location(),
	      name: 'Location'
	  },
	  {
	      viz: pileup.viz.coverage(),
	      data: bamSource,
	      name: 'Coverage'
	  },
	  {
	      viz: pileup.viz.pileup({
	        viewAsPairs: true
	      }),
	      data: bamSource,
	      cssClass: 'normal',
	      name: 'Alignments'
	  }
	];
window.p = pileup.create(divGB, {
		  range: range,
		  tracks: sources
		});
```

- **_T3:_** reference(hg38.2bit) from biodallience and .bam and .bai files from local file in the computer. The reference loads as before (in pieces and very fast) but the bam file weights 1,7 GB and it takes very long to load bringing the app to crush.

**JAVASCRIPT CODE:**
```javascript
window.divGB = document.getElementById('genomeBrowserHolder');
window.range = {contig: 'chr17', start: 7512384, stop: 7512544};
window.bamSource = pileup.formats.bam({
    	url: "{{MEDIA_URL}}uploads/{{project_id}}/bwa/myfile.bam",
        indexUrl: "{{MEDIA_URL}}uploads/{{project_id}}/bwa/myfile.bam.bai"
	});
window.sources = [
	  {
	      viz: pileup.viz.genome(),
	      isReference: true,
	      data: pileup.formats.twoBit({
	      	url: 'http://www.biodalliance.org/datasets/hg38.2bit'
	      }),
	      name: 'Reference'
	  },
	  {
	      viz: pileup.viz.scale(),
	      name: 'Scale'
	  },
	  {
	      viz: pileup.viz.location(),
	      name: 'Location'
	  },
	  {
	      viz: pileup.viz.coverage(),
	      data: bamSource,
	      name: 'Coverage'
	  },
	  {
	      viz: pileup.viz.pileup({
	        viewAsPairs: true
	      }),
	      data: bamSource,
	      cssClass: 'normal',
	      name: 'Alignments'
	  }
	];
window.p = pileup.create(divGB, {
		  range: range,
		  tracks: sources
		});
```
**Question:**

I have been trying so many things to work around the problem of this slow load when files are stored locally but I found no solution yet... I would expect a performance close to the one in the **_T1_** where resources are loaded whenever they are necessary and not before, so the pileup browser can render faster the results.

I would really appreciate if anyone could tell me what am I doing wrong and which could be a way to solve it.

Thank you very much

**Comment by armish** [comment]
Hi @ifll - sorry to hear about this performance issue. There should be something weird going on as there normally shouldn't be a difference between hosting files locally/remotely (assuming that the bandwidth is the same).

One immediate thing that comes to my mind is your local server setup: pileup.js heavily depends on RANGE requests to only work with relevant portions of the "remote" files, so can you please verify that your local server respects these requests and performs range requests fast enough?

If this is not the issue, I will try to debug what might be going wrong in your case.

Thanks!
-- Arman

**Comment by ifll** [comment]
Hi @armish 

Thanks for your fast reply. Yesterday I have been trying to deal with this issue and I activated the Django debug mode to check what was going wrong... but couldn't find it... So far, I must say I don't really know if it is possible to activate range requests within Django, so it might be that I didn't find the solution on that way.

Anyway I decided to go for a workaround solution which fixed this issue for me. I build up a local fileserver with Cherokee and allowed RANGE requests on it, then it works like charm as I was expecting.

Thank you very much for you help
